File: scripts/lib/mobile-phone-api.mjs
/**
 * 移动 phone 通用 API（Node）— back / goto，供 ada-fluent 与 MCP 适配器复用
 */
import { wait } from "./ada.mjs";
import { runLaunchSettle } from "./smart-wait-launch.mjs";

export function isAppBundleId(value) {
  return typeof value === "string" && /^[a-z][a-z0-9_]*(\.[a-z][a-z0-9_]+)+$/i.test(value);
}

/** @param {(cmd: string, extra?: object) => Promise<unknown>} run */
export function createBack(run) {
  return async (times = 1, gapMs = 400) => {
    const n = Math.max(1, Number(times) || 1);
    for (let i = 0; i < n; i++) {
      await run("back");
      if (i < n - 1) await wait(gapMs);
    }
  };
}

/**
 * @param {"android"|"ios"|"harmony"} platform
 * @param {(loc: unknown) => { exists: () => Promise<boolean>, click: () => Promise<void> }} find
 * @param {(cmd: string, extra?: object) => Promise<unknown>} run
 */
export function createGoto(platform, find, run) {
  return async (target, second, third) => {
    if (target == null) throw new Error("goto: target is required");

    if (typeof target === "object") {
      const { appId, bundleId, abilityId, ability, settleMs, wait: waitOpts } = target;
      const id = appId ?? bundleId;
      if (!id) throw new Error("goto: appId is required");
      const extra = { appId: id };
      if (platform === "harmony") extra.abilityId = abilityId ?? ability ?? "EntryAbility";
      await run("launchApp", extra);
      await runLaunchSettle(run, platform, settleMs, waitOpts);
      return;
    }

    if (typeof target === "string" && isAppBundleId(target)) {
      const extra = { appId: target };
      if (platform === "harmony") {
        extra.abilityId = typeof second === "string" ? second : "EntryAbility";
        const settleMs = typeof second === "number" ? second : typeof third === "number" ? third : undefined;
        await run("launchApp", extra);
        await runLaunchSettle(run, platform, settleMs);
        return;
      }
      const settleMs = typeof second === "number" ? second : undefined;
      await run("launchApp", extra);
      await runLaunchSettle(run, platform, settleMs);
      return;
    }

    const labels = Array.isArray(target) ? target : [target];
    for (const label of labels) {
      const handle = find(label);
      if (await handle.exists()) {
        await handle.click();
        return;
      }
    }
    throw new Error(`goto: page not found: ${JSON.stringify(target)}`);
  };
}

File: scripts/lib/mobile_phone_api.py
"""移动 phone 通用 API（Python）— back / goto。"""
from __future__ import annotations

import re
import time
from typing import Any, Callable

from smart_wait_launch import run_launch_settle


def is_app_bundle_id(value: str) -> bool:
    # 各段须以字母或数字开头（兼容 com.360buy.jdmobile 等数字段 bundle）
    return bool(re.match(r"^[a-zA-Z][a-zA-Z0-9_-]*(\.[a-zA-Z0-9][a-zA-Z0-9_-]*)+$", value, re.I))


def back_times(
    run: Callable[[str, dict[str, Any]], None], times: int = 1, gap_ms: float = 400
) -> None:
    n = max(1, times)
    for i in range(n):
        run("back", {})
        if i < n - 1:
            time.sleep(max(0.0, float(gap_ms)) / 1000.0)


def _settle_after_launch(
    run: Callable[[str, dict[str, Any]], None],
    platform: str,
    settle_ms: int | float | None,
    explicit_wait: dict[str, Any] | None = None,
) -> None:
    run_launch_settle(run, platform, settle_ms, explicit_wait)


def goto_target(
    platform: str,
    find: Callable[[Any], Any],
    run: Callable[[str, dict[str, Any]], None],
    target: str | list[str] | dict[str, Any],
    second: str | int | None = None,
    third: int | None = None,
) -> None:
    if isinstance(target, dict):
        app_id = target.get("appId") or target.get("bundleId")
        if not app_id:
            raise RuntimeError("goto: appId is required")
        extra: dict[str, Any] = {"appId": str(app_id)}
        if platform == "harmony":
            extra["abilityId"] = str(target.get("abilityId") or target.get("ability") or "EntryAbility")
        run("launchApp", extra)
        _settle_after_launch(
            run,
            platform,
            target.get("settleMs") if isinstance(target.get("settleMs"), (int, float)) else None,
            target.get("wait") if isinstance(target.get("wait"), dict) else None,
        )
        return

    if isinstance(target, str) and is_app_bundle_id(target):
        extra = {"appId": target}
        if platform == "harmony":
            ability = second if isinstance(second, str) else "EntryAbility"
            extra["abilityId"] = ability
            settle_ms = third if isinstance(third, (int, float)) else (second if isinstance(second, (int, float)) else None)
        else:
            settle_ms = second if isinstance(second, (int, float)) else None
        run("launchApp", extra)
        _settle_after_launch(run, platform, settle_ms)
        return

    labels = target if isinstance(target, list) else [target]
    for label in labels:
        el = find(str(label))
        if el.exists():
            el.click()
            return
    raise RuntimeError(f"goto: page not found: {target!r}")
